src/moustachupkg/wrapcontext.nim

```diff
@@ -0,0 +1,166 @@
+        else:
+          if seq_of.startsWith("seq["): # seq of seq; handled by Context
+            result.add("  result[\"$1\"] = nd.$1\n".format(fn))
+          elif seq_of in object_types_seen:
+            result.add("  result[\"$1\"] = map(nd.$1, proc(x: $2): Context = x.toContext())\n".format(fn, seq_of))
+          else:
+            echo "WARNING: skipping $1.$2 as $3 it is not recognized by wrap_context_methods".format(object_name, fn, ft)
+      elif ft in object_types_seen:
+        result.add("  result[\"$1\"] = nd.$1.toContext()\n".format(fn))
+      else:
+        echo "WARNING: skipping $1.$2 as $3 it is not recognized by wrap_context_methods".format(object_name, fn, ft)
+  result.add("\n")
+
+proc generate_apply_string(object_name: string, field_type_map: OrderedTable[string, string]): string {.compileTime.} =
+  result = ""
+  result.add("proc apply*(c: var Context, nd: $1) =\n".format(object_name))
+  for fn, ft in field_type_map:
+    case ft:
+    of "string", "int", "float", "bool":
+      result.add("  c[\"$1\"] = nd.$1\n".format(fn))
+    of "Oid":
+      result.add("  c[\"$1\"] = $$nd.$1\n".format(fn))
+    else:
+      if ft.startsWith("seq["):
+        var seq_of = ft[4 ..< ft.high]
+        case ft:
+        of "string", "int", "float", "bool":
+          result.add("  c[\"$1\"] = nd.$1\n".format(fn))
+        of "Oid":
+          result.add("  c[\"$1\"] = map(nd.$1, proc(x: Oid): string = $$x)\n".format(fn, seq_of))
+        else:
+          if seq_of.startsWith("seq["): # seq of seq; handled by Context
+            result.add("  c[\"$1\"] = nd.$1\n".format(fn))
+          elif seq_of in object_types_seen:
+            result.add("  c[\"$1\"] = map(nd.$1, proc(x: $2): Context = x.toContext())\n".format(fn, seq_of))
+          else:
+            echo "WARNING: skipping $1.$2 as $3 it is not recognized by wrap_context_methods".format(object_name, fn, ft)
+      elif ft in object_types_seen:
+        result.add("  c[\"$1\"] = nd.$1.toContext()\n".format(fn))
+      else:
+        echo "WARNING: skipping $1.$2 as $3 it is not recognized by wrap_context_methods".format(object_name, fn, ft)
+  result.add("\n")
+
+
+# ######################################
+#
+# MACROS
+#
+# ######################################
+
+macro wrap_context*(n: typed): typed =
+  var s: string = ""
+  #
+  # gather details about n (mongo_object) object
+  #
+  let object_name = $n
+  echo "context object \"$1\" added".format(object_name)
+  object_types_seen.add(object_name)
+  var field_type_map = get_fields(n)
+  var field_names: seq[string] = @[]
+  var field_types: seq[string] = @[]
+  for fn, ft in field_type_map.pairs():
+    field_names.add(fn)
+    if not field_types.contains(ft):
+      field_types.add(ft)
+  #
+  # toContext()
+  #
+  s.add(generate_toContext_string(object_name, field_type_map))
+  #
+  # context.apply(obj)
+  #
+  s.add(generate_apply_string(object_name, field_type_map))
+  #
+  # generate it all
+  #
+  parseStmt s
+
```

Comment:
I won't be taking this file in. I can see its usefulness for object rendering. However, its purpose is more specific than the library wants to be and would fit better in a supporting web framework library (that seems to be the actual goal anyway, right?).